apps/dashboard/app/models/project.rb

```diff
@@ -6,19 +6,42 @@ class Project
   include ActiveModel::Validations
 
   class << self
-    def all
-      return [] unless dataroot.directory? && dataroot.executable? && dataroot.readable?
+    def lookup_file
+      Pathname("#{dataroot}/.project_lookup").tap do |path|
+        FileUtils.touch(path.to_s) unless path.exist?
+      end
+    end
 
-      dataroot.children.map do |d|
-        Project.new({ :name => d.basename })
-      rescue StandardError => e
-        Rails.logger.warn("Didn't create project. #{e.message}")
-        nil
-      end.compact
+    def lookup_table
+      f = File.read(lookup_file)
+      YAML.safe_load(f).to_h
+    rescue StandardError, Exception => e
+      Rails.logger.warn("cannot read #{dataroot}/.project_lookup due to error #{e}")
+      {}
     end
 
-    def find(project_pathname)
-      from_directory(dataroot.join(project_pathname))
+    def next_id
+      lookup_table.map { |id, _directory| id }
+                  .map(&:to_i)
+                  .concat([0]) # could be the first project
+                  .max + 1
+    end
+
+    def all
+      lookup_table.map do |id, directory|
+        Project.new({ id: id, directory: directory })
+      end
+    end
+
+    def find(id)
+      opts = lookup_table.select do |lookup_id, _directory|
+        lookup_id == id.to_i
+      end.map do |lookup_id, directory|
+        { id: lookup_id, directory: directory }
+      end.first
+      return nil if opts.nil?
+
+      Project.new(opts)
```

Comment:
```suggestion
    def find(id)
      return nil if opts.nil?
       
      opts = lookup_table.select do |lookup_id, _directory|
        lookup_id == id.to_i
      end.map do |lookup_id, directory|
        { id: lookup_id, directory: directory }
      end.first

      Project.new(opts)
```
Might be better to have a guard here instead of at the end.

Reply:
`opts` isn't defined at that point.  The way you have it, we'd get an `undefined local variable or method 'opts'` on line 37 because it's not defined until line 39.